### PR TITLE
Sync adherent data with Mailchimp

### PR DESCRIPTION
**Pull request by ottaviano**

### CODE REVIEW

File: app/config/services/messenger.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        - { id: send_message }
+        - { id: handle_message }
+
+services:
+    _defaults:
+        autoconfigure: true
+        autowire: true
+        public: false
+
+    # Message Bus with it alias
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply app/config/services/messenger.yml]: @JeremieSamson thanks 👍

File: src/Mailchimp/Synchronisation/RequestBuilder.php [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            ->setCity($adherent->getCityName())
+            ->setInterests(
+                array_replace(
+                    array_fill_keys($interestIds, false),
+                    array_fill_keys(
+                        array_intersect_key($interestIds, array_flip($adherent->getInterests())),
+                        true
+                    ),
+                    /*
+                     * This is a hack for to migrate progressively to the ID stored
```
Comment (JeremieSamson) [inline src/Mailchimp/Synchronisation/RequestBuilder.php]:
```suggestion
                     * This is a hack to migrate progressively the ID stored
```

File: src/Mailchimp/Synchronisation/RequestBuilder.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                     * into DB (subscription_types.external_id column), after that we will can use this method:
```
Comment (JeremieSamson) [inline src/Mailchimp/Synchronisation/RequestBuilder.php]:
```suggestion
                     * into DB (subscription_types.external_id column), after that we will be able to use this method:
```

File: src/Mailchimp/Synchronisation/RequestBuilder.php [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    public function createMemberTagsRequest(string $memberIdentifier, array $removedTags = []): MemberTagsRequest
+    {
+        $request = new MemberTagsRequest($memberIdentifier);
+
+        foreach (array_merge($removedTags, $this->inactiveTags) as $tagName) {
+            $request->addTag($tagName, false);
+        }
+
+        foreach ($this->activeTags as $tagName) {
+            $request->addTag($tagName, true);
```
Comment (JeremieSamson) [inline src/Mailchimp/Synchronisation/RequestBuilder.php]: `public function addTag(string $tagName, bool $active = true)`
active id true by default, was it intentional ?

File: src/Controller/EnMarche/MembershipController.php [hunk 5]
```diff
@@ -189,14 +189,18 @@ public function completeAction(MembershipRegistrationProcess $membershipRegistra
...
+        CallbackManager $callbackManager,
+        AdherentAccountActivationHandler $accountActivationHandler
+    ): Response {
         if ($this->getUser()) {
             $this->redirectToRoute('app_search_events');
         }
 
         try {
-            $this->get('app.adherent_account_activation_handler')->handle($adherent, $activationToken);
+            $accountActivationHandler->handle($adherent, $activationToken);
```
Comment (JeremieSamson) [inline src/Controller/EnMarche/MembershipController.php]: I am wondering why are you injecting `AdherentAccountActivationHandler` here but not `MembershipRequestHandler::class` 3 lines above ?

Comment (ottaviano) [reply src/Controller/EnMarche/MembershipController.php]: I have modified `AdherentAccountActivationHandler` service by adding a new argument and I make it private. As this service is now private, I change **only** its injections and don't touch `MembershipRequestHandler` service for staying in the scope of my card.

File: src/Mailchimp/Synchronisation/Driver.php [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        return $this->sendRequest(
+            'PUT',
+            sprintf('/lists/%s/members/%s', $this->listId, md5($request->getMemberIdentifier())),
+            $request->toArray()
+        );
+    }
+
+    /**
+     * Update member's tags
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Mailchimp/Synchronisation/Driver.php]: removed

File: src/Repository/CommitteeMembershipRepository.php [hunk 7]
```diff
@@ -251,6 +251,8 @@ public function findMembers(Committee $committee): AdherentCollection
      * that does not already belongs to list of all members of $destinationCommittee
      *
      * NOTE: this returns poorly hydrated instances of Adherent, but is optimized for merging of committees
+     *
+     * @return Adherent[]|self
```
Comment (illusionOfParadise) [inline src/Repository/CommitteeMembershipRepository.php]: Return doc seems not good: this method never returns `self`.

Comment (ottaviano) [reply src/Repository/CommitteeMembershipRepository.php]: fixed, thanks

File: src/Mailchimp/Synchronisation/Manager.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        );
+
+        if ($result) {
+            // Active/Inactive member's tags
+            $result = $this->driver->updateMemberTags(
+                $requestBuilder->createMemberTagsRequest($message->getEmailAddress(), $message->getRemovedTags())
+            );
+
+            if ($result) {
+                $this->logger->info(sprintf('Mailchimp member "%s" has been updated', $adherent->getUuidAsString()));
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Mailchimp/Synchronisation/Manager.php]: I have added a new log for `member edition`, when edition can be a creation or modification

File: src/Committee/CommitteeManager.php [hunk 9]
```diff
@@ -369,9 +373,9 @@ private function doUnfollowCommittee(CommitteeMembership $membership, Committee
 
         $manager->persist($this->createCommitteeMembershipHistory($membership, CommitteeMembershipAction::LEAVE()));
 
-        if ($flush) {
-            $manager->flush();
-        }
+        $manager->flush();
+
+        $this->dispatcher->dispatch(UserEvents::USER_UPDATE_COMMITTEE_PRIVILEGE, new UserEvent($membership->getAdherent()));
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Committee/CommitteeManager.php]: done

File: src/Committee/CommitteeManager.php [hunk 10]
```diff
@@ -491,4 +472,36 @@ private function createCommitteeMembershipHistory(CommitteeMembership $membershi
     {
         return new CommitteeMembershipHistory($membership, $action);
     }
+
+    private function changePrivilegeOnMembership(CommitteeMembership $membership, string $privilege, bool $flush = true): void
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Committee/CommitteeManager.php]: done

File: src/Mailchimp/Synchronisation/EventListener/AdherentEventSubscriber.php [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private function dispatchMessage(UuidInterface $uuid, string $identifier, array $removedTags = []): void
+    {
+        $this->bus->dispatch(new AdherentChangeMessage($uuid, $identifier, $removedTags));
+    }
+
+    private function transformToArray(Adherent $adherent): array
+    {
+        return $this->normalizer->toArray($adherent, SerializationContext::create()->setGroups(['change_diff']));
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Mailchimp/Synchronisation/EventListener/AdherentEventSubscriber.php]: replaced by `adherent_change_diff`, thanks

File: src/Mailchimp/Synchronisation/Manager.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (thomasbisignani) [reply src/Mailchimp/Synchronisation/Manager.php]: Nice, thx :+1:

**Pull request merged**